### PR TITLE
Authorization header casing

### PR DESCRIPTION
**Pull request by ojongerius**
I'd like to support "Authorization" which would comply with https://tools.ietf.org/html/rfc7235#section-4.2

### CODE REVIEW

**Comment by Bouncey** [comment]
In all my testing for #58 I only ever changed the token, the `Authorization` field never changed and I always got `headers.authorization` at the directive. Apart from placing the headers directly on to context, we do not manipulate them. Do the directives still work with these changes?

I screenshot of the headers I used during #58 
screenshot_20180430-071436_firefox

**Comment by Bouncey** [comment]
Further to above, I have just logged out what we have in the headers in `handler`

```js
// handler.js
// ...
log(Object.keys(headers));
// ...
```
```sh
  fcc:handler [ 'Host',
  fcc:handler   'User-Agent',
  fcc:handler   'Accept',
  fcc:handler   'Accept-Language',
  fcc:handler   'Accept-Encoding',
  fcc:handler   'Referer',
  fcc:handler   'content-type',
  fcc:handler   'authorization',
  fcc:handler   'origin',
  fcc:handler   'Content-Length',
  fcc:handler   'Connection',
  fcc:handler   'Content-Type' ]
```

File: auth/index.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ const { JWT_CERT } = process.env;
 
 export function verifyWebToken(ctx) {
   log('Verifying token');
-  const token = ctx && ctx.headers && ctx.headers.authorization;
+  const token = ctx && ctx.headers && ctx.headers.Authorization;
```
Comment (Bouncey) [inline auth/index.js]: This change will mean that `token` will always be `false` as `Authorization` is not available in `headers`

**Review by Bouncey — CHANGES_REQUESTED** [review]

**Comment by raisedadead** [comment]
```
...
  fcc:handler   'content-type', <----------------
  fcc:handler   'authorization',
  fcc:handler   'origin',
  fcc:handler   'Content-Length',
  fcc:handler   'Connection',
  fcc:handler   'Content-Type' ] <---------------

```

There are duplicates for the same keys, this leads me to think that this is being set by someone somewhere, (could be programmatic, by some middle-ware or even us in our logic), which needs investigation.

File: auth/index.js (hunk 1 above)
Comment (raisedadead) [reply auth/index.js]: Can be mitigated with a `ctx && ctx.headers && (ctx.headers.Authorization || ctx.headers.authorization)`.

But I would like to know how is this being set in the first place.

**Comment by ojongerius** [comment]
Interesting 😄 

Running against staging f3f971cf6a075bbd1d0bc862838463d2cc0bb45e:

Using "Authorization":

```sh
▶ curl -s -X POST -H "Content-Type: application/json" --data '{ "query": "{ users { _id name email } }" }' localhost:4000/graphql --header "Authorization: Bearer ${JWT}"  |jq
{
  "data": {
    "users": null
  },
  "errors": [
    {
      "message": "You must supply a JSON Web Token for authorization!",
      "locations": [
        {
          "line": 1,
          "column": 3
        }
      ],
      "path": [
        "users"
      ]
    }
  ]
}
```

Using "authorization":

```sh
▶ curl -s -X POST -H "Content-Type: application/json" --data '{ "query": "{ users { _id name email } }" }' localhost:4000/graphql --header "authorization: Bearer ${JWT}"  |jq
{
  "data": {
    "users": [
      {
        "_id": "5acd3a7e9f80cf00248abc97",
        "name": "",
        "email": "otto@example.com"
      },
      {
        "_id": "5ad1c7ace769e064e3a3487b",
        "name": "Yet another Otto",
        "email": "otto@a20.net"
      },
      {
        "_id": "5ad1c7eee769e01038a3487c",
        "name": "Frida Kahlo",
        "email": "frida@example.com"
      },
      {
        "_id": "5ad5392f2109c9843e5939d7",
        "name": "Karel Appel",
        "email": "karel@example.com"
      },
      {
        "_id": "5ad5817bb8d47b9b06e2d0f2",
        "name": "Quincy Larson",
        "email": "quincy@example.com"
      },
      {
        "_id": "5ad6aa28948979c90662ef4a",
        "name": "Quincy Larson",
        "email": "quincy@example.com"
      },
      {
        "_id": "5ad6acae94897931e962ef4b",
        "name": "Quincy Larson",
        "email": "quincy@example.com"
      }
    ]
  }
}
```

@Bouncey what browser are you using?

So the HTTP/2 specs says that headers "header field names MUST be converted to lowercase prior to their encoding in HTTP/2". 

And the RFC actually specifies that " Field name are case-insensitive." (https://tools.ietf.org/html/rfc2616#section-4.2).

File: auth/index.js (hunk 1 above)
Comment (ojongerius) [reply auth/index.js]: Reading a bit more about this is sounds like we do actually need to support upper and lowercase for header field names.

**Comment by ojongerius** [comment]
Can I summarise that we all agree that, if a client sets some headers, in this case "Authorization" with the first letter being an uppercase the behaviour is incorrect?

I'm surprised that Stuarts test works initially and am wondering if his client lowercased it for him. I suspect this to be the case, I found https://github.com/graphcool/graphql-playground/issues/370 which sound similar, but I stumbled upon this bug while trying to use `Authorization` instead of `authorization` in the playground served at /api 

@raisedadead I think you were spot in your suspicions: there is an interesting interaction between the HAPI framework used by serverless-offline which at leased used to lowercase header field names.

To make it more interesting, the Hapi framework used by serverless-offline (ie: running it locally) will lowercase header field names, but when you deploy it to AWS headers will come in as is -I've not verified this but base it on comment https://github.com/dherault/serverless-offline/issues/260#issuecomment-331037948.

For now I'd suggest that we use `ctx && ctx.headers && (ctx.headers.Authorization || ctx.headers.authorization)`.

File: auth/index.js [hunk 2]
```diff
@@ -8,7 +8,10 @@ const { JWT_CERT } = process.env;
 
 export function verifyWebToken(ctx) {
   log('Verifying token');
-  const token = ctx && ctx.headers && ctx.headers.authorization;
+  const token =
+    ctx &&
+    ctx.headers &&
+    (ctx.headers.Authorization || ctx.headers.authorization);
```
Comment (ojongerius) [inline auth/index.js]: But what do we do if both "Authorization" and "authorization" are present? I'm not sure, the spec says to lowercase the headers field names, which we could do but we'd still need to decide what to do with duplicates. Open to suggestions.

Comment (raisedadead) [reply auth/index.js]: For duplicates, I think its fair to stick to the lower-case values. And as of now, we being the only-consumers need to take care of this client side too.

Comment (raisedadead) [reply auth/index.js]: In that case I would pass all the headers though a util and turn them all to lowercase.

Comment (ojongerius) [reply auth/index.js]: I'd rather not mess with the headers manually, and have us support both mix and all lower. I've changed it to prefer all lower, which would be what most clients would set and many libraries normalise to anyway.

**Comment by Bouncey** [comment]
## LGTM

@raisedadead are you happy with the work-around?

**Review by Bouncey — APPROVED** [review]

**Comment by ojongerius** [comment]
@raisedadead good to go?

**Review by raisedadead — APPROVED** [review]

**Pull request merged**